catlearn/graph_utils.py

```diff
@@ -536,3 +540,129 @@ def generate_random_graph(
+
+
+def sample(
+        igraph: DirectedGraph[NodeType],
+        sample_vertices_size: int,
+        ranking: Callable[[DirectedGraph[NodeType]], Mapping[NodeType, float]],
+        rng: random.Random) -> DirectedGraph[NodeType]:
+    """
+    Sample a random subgraph of `igraph` with respect to a probability
+    distribution `ranking` over the vertices.
+
+    Params:
+    - igraph: Input graph
+    - sample_vertices_size: number of vertices to sample
+        (with replacement so actual number might be lower)
+    - ranking: probability distribution over the input graph vertices
+    - rng: Random generator
+
+    Returns:
+    A valid random subgraph
+
+    """
+    ranks = list(ranking(igraph).items())
+    vertices = [v for v, _ in ranks]
+    weights = [w for _, w in ranks]
+    sampled_vertices = set(rng.choices(vertices, weights, k=int(sample_vertices_size)))
+    return igraph.subgraph(sampled_vertices)
+
+
+def pagerank_sample(
+        igraph: DirectedGraph[NodeType],
+        sample_vertices_size: int,
+        rng: random.Random,
+        **kwargs) -> DirectedGraph[NodeType]:
+    """
+    Sample a random subgraph of `igraph` with respect to vertices PageRank scores.
+
+    Details of other parameters: see `sample`
+    Pagerank calculation: see `networkx.pagerank`
+    NB: `kwargs are all passed to `networkx.pagerank`
+    """
+    return sample(
+        igraph, sample_vertices_size, lambda g: pagerank(g, **kwargs), rng)
+
+
+def hubs_sample(
+        igraph: DirectedGraph[NodeType],
+        sample_vertices_size: int,
+        rng: random.Random,
+        **kwargs) -> DirectedGraph[NodeType]:
+    """
+    Sample a random subgraph of `igraph` with respect to vertices HITS hubs scores.
+
+    Details of other parameters: see `sample`
+    Pagerank calculation: see `networkx.hits`
+    NB: `kwargs are all passed to `networkx.hits`
+    """
+    return sample(
+        igraph, sample_vertices_size, lambda g: hits(g, **kwargs)[0], rng)
+
+
+def authorities_sample(
+        igraph: DirectedGraph[NodeType],
+        sample_vertices_size: int,
+        rng: random.Random,
+        **kwargs) -> DirectedGraph[NodeType]:
+    """
+    Sample a random subgraph of `igraph` with respect to vertices HITS authorities scores.
+
+    Details of other parameters: see `sample`
+    Pagerank calculation: see `networkx.hits`
+    NB: `kwargs are all passed to `networkx.hits`
+    """
+    return sample(
+        igraph, sample_vertices_size, lambda g: hits(g, **kwargs)[1], rng)
+
+
+def uniform_sample(
+        igraph: DirectedGraph[NodeType],
+        sample_vertices_size: int,
+        rng: random.Random) -> DirectedGraph[NodeType]:
+    """
+    Sample a random subgraph of `igraph` with a uniform probability over vertices
+
+    Details of parameters: see `sample`
+    """
+    n = len(igraph)
+    return sample(igraph, sample_vertices_size,
+                  lambda G: {v: 1.0/n for v in G}, rng)
+
+def random_walk_sample(
```

Comment:
Open question: Is the current function really what we want?
The way I see it there is 2 options:
- the current one, where we keep only the nodes which have been sampled
- the one in which instead we keep the edge which have been used during the walk (dunno if that's even practical using networkx). (obviously we also only keep the visited nodes)